src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/VisitServiceTest.kt

```diff
@@ -0,0 +1,181 @@
+
+    whenever(offenderVisitBalanceRepository.findById(offenderBookingId)).thenReturn(
+      Optional.of(
+        OffenderVisitBalance(
+          offenderBookingId = offenderBookingId,
+          remainingVisitOrders = 3,
+          remainingPrivilegedVisitOrders = 5,
+        )
+      )
+    )
+  }
+
+  @Nested
+  internal inner class createVisit {
+    @Test
+    fun success() {
+
+      whenever(visitRepository.save(any())).thenReturn(Visit(id = visitId))
+
+      assertThat(visitService.createVisit(createVisitRequest)).isEqualTo(CreateVisitResponse(visitId))
+
+      val balanceArgumentCaptor = ArgumentCaptor.forClass(OffenderVisitBalanceAdjustment::class.java)
+      verify(offenderVisitBalanceAdjustmentRepository).save(balanceArgumentCaptor.capture())
+      val balanceArgument = balanceArgumentCaptor.value
+      assertThat(balanceArgument.adjustReasonCode).isEqualTo(VisitOrderAdjustmentReason.VISIT_ORDER_ISSUE)
+      assertThat(balanceArgument.remainingVisitOrders).isEqualTo(-1)
+      assertThat(balanceArgument.remainingPrivilegedVisitOrders).isNull()
+      assertThat(balanceArgument.commentText).isEqualTo("Created by PVB3 for an on-line visit booking")
+
+      val visitArgumentCaptor = ArgumentCaptor.forClass(Visit::class.java)
+      verify(visitRepository).save(visitArgumentCaptor.capture())
+      val visit = visitArgumentCaptor.value
+      assertThat(visit.visitDate).isEqualTo(LocalDate.of(2021, 11, 4))
+      assertThat(visit.startTime).isEqualTo(LocalDateTime.of(2021, 11, 4, 12, 5))
+      assertThat(visit.endTime).isEqualTo(LocalDateTime.of(2021, 11, 4, 13, 4))
+      assertThat(visit.visitType).isEqualTo(visitType)
+      assertThat(visit.visitStatus).isEqualTo(visitStatus)
+      assertThat(visit.location?.id).isEqualTo(prisonId)
+
+      val inOrder = Mockito.inOrder(visitVisitorRepository)
+      val visitorArgumentCaptor = ArgumentCaptor.forClass(VisitVisitor::class.java)
+      inOrder.verify(visitVisitorRepository, times(3)).save(visitorArgumentCaptor.capture())
+      assertThat(visitorArgumentCaptor.allValues).extracting(
+        "visitId", "offenderBooking.bookingId", "personId", "eventStatus"
+      ).containsExactly(
+        Tuple.tuple(visitId, offenderBookingId, null, eventStatus),
+        Tuple.tuple(visitId, null, 45L, eventStatus),
+        Tuple.tuple(visitId, null, 46L, eventStatus),
+      )
+    }
+
+    @Test
+    fun successWithPrivilege() {
+
+      whenever(visitRepository.save(any())).thenReturn(Visit(id = visitId))
+
+      assertThat(visitService.createVisit(createVisitRequest.copy(privileged = true))).isEqualTo(CreateVisitResponse(visitId))
+
+      val balanceArgumentCaptor = ArgumentCaptor.forClass(OffenderVisitBalanceAdjustment::class.java)
+      verify(offenderVisitBalanceAdjustmentRepository).save(balanceArgumentCaptor.capture())
+      val balanceArgument = balanceArgumentCaptor.value
```

Comment:
you could probably use `check {}`  and assertj rather than an ArgumentCaptor

Reply:
thanks, yes its much clearer (except I dont think I can use it in the inOrder case)